File: frontend-v2/src/shared/api/auth-store.ts
// Auth token state. Access token lives in memory only; the refresh token is
// an HttpOnly cookie owned by the backend (ported from v1 — keep it this way).
import { create } from "zustand"
import { env } from "@/shared/config/env"
import type { AuthUser } from "@/shared/types/api"

interface AuthState {
  accessToken: string | null
  user: AuthUser | null
  isAuthenticated: boolean
  isLoading: boolean
  setAuth: (token: string, user: AuthUser | null) => void
  clearAuth: () => void
  setLoading: (loading: boolean) => void
}

export const useAuthStore = create<AuthState>((set) => ({
  accessToken: null,
  user: null,
  isAuthenticated: false,
  isLoading: true, // stays true until the first refresh attempt settles
  setAuth: (token, user) => set({ accessToken: token, user, isAuthenticated: true, isLoading: false }),
  clearAuth: () => set({ accessToken: null, user: null, isAuthenticated: false, isLoading: false }),
  setLoading: (loading) => set({ isLoading: loading }),
}))

// ── Refresh with mutex: concurrent 401s trigger exactly one refresh ──
let refreshPromise: Promise<string | null> | null = null
const SINGLE_USER_ACCESS_TOKEN = "openbox-single-user"

export function refreshAccessToken(): Promise<string | null> {
  refreshPromise ??= doRefresh().finally(() => {
    refreshPromise = null
  })
  return refreshPromise
}

async function doRefresh(): Promise<string | null> {
  try {
    const resp = await fetch(`${env.apiBase}/api/auth/refresh`, {
      method: "POST",
      credentials: "include",
    })
    if (!resp.ok) {
      // Desktop/single-user deployments intentionally omit all credential
      // routes, so /refresh is a 404 rather than a missing-cookie 401.  The
      // public bootstrap contract is authoritative and returns user data only
      // when authentication is disabled; SaaS deployments report multi_user.
      const bootstrapResp = await fetch(`${env.apiBase}/api/auth/bootstrap`, {
        credentials: "include",
      })
      if (bootstrapResp.ok) {
        const bootstrap = (await bootstrapResp.json()) as {
          mode: "single_user" | "multi_user"
          user?: AuthUser
        }
        if (bootstrap.mode === "single_user" && bootstrap.user) {
          const user = bootstrap.user
          useAuthStore.getState().setAuth(SINGLE_USER_ACCESS_TOKEN, user)
          return SINGLE_USER_ACCESS_TOKEN
        }
      }
      useAuthStore.getState().clearAuth()
      return null
    }
    const data = (await resp.json()) as { access_token: string }
    const token = data.access_token
    const meResp = await fetch(`${env.apiBase}/api/auth/me`, {
      headers: { Authorization: `Bearer ${token}` },
    })
    const user = meResp.ok ? ((await meResp.json()) as AuthUser) : null
    useAuthStore.getState().setAuth(token, user)
    return token
  } catch {
    useAuthStore.getState().clearAuth()
    return null
  }
}

File: mobile/lib/shared/api/auth_store.dart
import 'dart:async';

import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../i18n/i18n.dart';
import '../models/auth_user.dart';
import '../models/json.dart';
import '../notifications/system_notifications.dart';
import '../widgets/toast.dart';
import '../ws/ws_client.dart';
import 'api_error.dart';
import 'logto_session.dart';
import 'providers.dart';

/// Global auth store, mirroring frontend-v2 `shared/api/auth-store.ts`:
/// access token in memory only (on [AuthSession]); refresh token rides the
/// cookie jar; `isLoading` is true until the boot refresh settles.
class AuthState {
  const AuthState({this.user, this.isLoading = true, this.mobileSessionId});

  final String? mobileSessionId;

  final AuthUser? user;
  final bool isLoading;

  bool get isAuthenticated => user != null;

  String get userId => user?.id ?? 'anonymous';
}

class AuthController extends Notifier<AuthState> {
  @override
  AuthState build() {
    ref.read(authSessionProvider).refreshFn = _doRefresh;
    ref.read(authSessionProvider).invalidateFn = invalidateMobileSession;
    return const AuthState();
  }

  /// Boot-time silent sign-in (web `refreshAccessToken()` before first
  /// paint): tries the refresh cookie; settles `isLoading`.
  Future<void> bootstrap() async {
    await ref.read(authSessionProvider).refresh();
    if (state.isLoading) {
      state = AuthState(
        user: state.user,
        mobileSessionId: state.mobileSessionId,
        isLoading: false,
      );
    }
  }

  void setAuth(String accessToken, AuthUser user, {String? mobileSessionId}) {
    final session = ref.read(authSessionProvider);
    session.revision++;
    session.mobileSessionId = mobileSessionId;
    session.accessToken = accessToken;
    session.userId = user.id;
    state = AuthState(
      user: user,
      mobileSessionId: mobileSessionId,
      isLoading: false,
    );
  }

  void clearAuth() {
    final session = ref.read(authSessionProvider);
    session.revision++;
    session.mobileSessionId = null;
    session.accessToken = null;
    session.userId = null;
    ref.read(wsClientProvider).disconnect();
    ref.read(workspaceScopeProvider).clear();
    state = const AuthState(isLoading: false);
  }

  void invalidateMobileSession(String code) {
    final hadUser = state.isAuthenticated;
    clearAuth();
    unawaited(ref.read(cookieJarProvider).deleteAll());
    unawaited(clearLocalLogtoTokens().catchError((Object _) {}));
    unawaited(ref.read(systemNotificationsProvider).clear());
    if (hadUser) {
      ref
          .read(toastProvider.notifier)
          .push(ToastKind.warning, ref.read(i18nProvider).t('errors:$code'));
    }
  }

  /// Sign out of both OpenBox and Logto. OpenBox state is always cleared even
  /// if either server is unavailable; Logto's SDK also removes native tokens
  /// in its own best-effort cleanup path.
  Future<void> signOut() async {
    final logtoConfig = ref.read(logtoSsoProvider.future);
    try {
      await ref.read(apiDioProvider).post<dynamic>('/api/auth/logout');
    } catch (_) {
      // Sign out locally regardless.
    }
    ref.read(wsClientProvider).disconnect();
    clearAuth();
    await ref.read(cookieJarProvider).deleteAll();
    unawaited(ref.read(systemNotificationsProvider).clear());
    try {
      await ref.read(logtoSessionProvider).signOut(await logtoConfig);
    } catch (_) {
      // The local OpenBox and Logto token state has already been cleared. A
      // canceled/unreachable browser logout must never restore this session.
    }
  }

  /// POST /api/auth/refresh (cookie) → GET /api/auth/me. Returns the new
  /// access token or null (→ signed out). Wired into [AuthSession] so
  /// concurrent 401s share one round-trip.
  Future<String?> _doRefresh() async {
    final dio = ref.read(refreshDioProvider);
    final revision = ref.read(authSessionProvider).revision;
    try {
      final refreshResp = await dio.post<Map<String, dynamic>>(
        '/api/auth/refresh',
      );
      if (revision != ref.read(authSessionProvider).revision) return null;
      final token = asString(refreshResp.data?['access_token']);
      if (token == null) {
        clearAuth();
        return null;
      }
      final meResp = await dio.get<Map<String, dynamic>>(
        '/api/auth/me',
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );
      final authSession = ref.read(authSessionProvider);
      if (revision != authSession.revision) return null;
      authSession.accessToken = token;
      authSession.mobileSessionId = asString(
        refreshResp.data?['mobile_session_id'],
      );
      final user = AuthUser.fromJson(meResp.data ?? const {});
      authSession.userId = user.id;
      state = AuthState(
        user: user,
        mobileSessionId: authSession.mobileSessionId,
        isLoading: false,
      );
      return token;
    } on DioException catch (error) {
      if (revision != ref.read(authSessionProvider).revision) return null;
      final code = ApiError.fromDio(error).code;
      if (code == 'AUTH_MOBILE_SESSION_REPLACED' ||
          code == 'AUTH_MOBILE_LOGIN_REQUIRED') {
        invalidateMobileSession(code);
      } else {
        clearAuth();
      }
      return null;
    }
  }
}

final authProvider = NotifierProvider<AuthController, AuthState>(
  AuthController.new,
);
